File: aspen_fota/client/src/com/htc/client/utils/AspenServiceModel.java
package com.htc.client.utils;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.RemoteException;
import android.util.Log;

import com.htc.service.IDeviceConnectedListener;
import com.htc.service.OnFirmwareUpdateListener;
import com.htc.service.IFotaService;

import java.util.concurrent.Executors;


/**
 * Created by hubin_jiang on 2018/6/18.
 */

public class AspenServiceModel {

    private static final String TAG = "AspenFota.client";

    public static final String SERVICE_PACKAGE = "com.htc.service";
    private static final String SERVICE_CLASS = SERVICE_PACKAGE + ".FotaService";

    private IFotaService mAspenService;
    private Handler mUIHandler = new Handler(Looper.getMainLooper());
    private myDelegate mDelegate;
    private boolean isBinded = false;

    private Context mContext;

    public AspenServiceModel(Context context) {
        mContext = context;
    }

    public boolean bind_FotaService() {
        try {
            Log.i(TAG, "__jh__ aidl bind_FotaService");
            Intent intent = new Intent();
            intent.setClassName(SERVICE_PACKAGE, SERVICE_CLASS);
            boolean isBind = mContext.bindService(intent, mConnection, Context.BIND_AUTO_CREATE);
            Log.d(TAG, "__jh__ isBind="+isBind);
            return isBind;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public void unbindService() {
        Log.d(TAG, "unbindService");
        mContext.unbindService(mConnection);
        isBinded = false;
    }

    public void setDelegate(myDelegate delegate) {
        mDelegate = delegate;
    }

    public myDelegate getDelegate() {
        return mDelegate;
    }

    private ServiceConnection mConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.d(TAG, "__jh__ onServiceConnected");
            isBinded = true;
            mAspenService = IFotaService.Stub.asInterface(service);
            try {
                mAspenService.setFirmwareUpdateListener(mFinchFotaListener);

                mAspenService.setDeviceConnectedListener(mAspenConnectLister);
            } catch (RemoteException e) {
                e.printStackTrace();
            }
            myDelegate delegate = getDelegate();
            if (delegate != null) {
                delegate.onServiceConnected();
            } else {
                Log.w(TAG, "[onServiceConnected] delegate is null !");
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };


    private IDeviceConnectedListener mAspenConnectLister =new IDeviceConnectedListener.Stub(){
        @Override
        public void onConnectedStateStatusChanged(int device, boolean isConnected, int usb_state) throws RemoteException {
            mUIHandler.post(new Runnable() {
                @Override
                public void run() {
                    myDelegate delegate = getDelegate();
                    if (delegate != null) {
                        //delegate.onServiceConnected();   //// todo
                    }
                }
            });
        }
    };

    private OnFirmwareUpdateListener mFinchFotaListener = new OnFirmwareUpdateListener.Stub() {
        @Override
        public void onFirmwareUpdateStatusChanged(int device, int state, Bundle extra) throws RemoteException {
            Log.d(TAG, "state : " + state);
            mUIHandler.post(new Runnable() {
                @Override
                public void run() {
                    myDelegate delegate = getDelegate();
                    if (delegate != null) {
                        //delegate.onDeviceStatusChanged(state, extra);   //// todo
                    }
                }
            });
        }
        @Override
        public void onFirmwareUpdateProgressChanged(int device, final int progress) throws RemoteException {
            mUIHandler.post(new Runnable() {
                @Override
                public void run() {
                    myDelegate delegate = getDelegate();
                    if (delegate != null) {
                        delegate.onFirmwareUpdateProgressChanged(progress);  //// todo
                    }
                }
            });
        }
    };


    @SuppressLint("StaticFieldLeak")
    public void getDeviceInfo() {
        new AsyncTask<Void, Void, Bundle>() {

            @Override
            protected Bundle doInBackground(Void... voids) {
                try {
                    return mAspenService.getDeviceInfo(FotaServiceContract.TYPE_DEVICE_HMD); //TODO
                } catch (Exception e) {
                    Log.e(TAG, "Exception on getDeviceInfo", e);
                }
                return null;
            }

            @Override
            protected void onPostExecute(Bundle info) {
                myDelegate delegate = getDelegate();
                if (delegate != null) {
                    if (info != null) {
                        delegate.onDeviceInfoGet(info);
                    } else {
                        delegate.onFotaError();
                    }
                } else {
                    Log.e(TAG, "delegate is null");
                }
            }
        }.executeOnExecutor(Executors.newFixedThreadPool(1));
    }

    public int getBatteryVoltageLevel() {
        int batteryLevel = -1;
        try {
            batteryLevel = mAspenService.getBatteryVoltageLevel(FotaServiceContract.TYPE_DEVICE_HMD);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return batteryLevel;
    }



    @SuppressLint("StaticFieldLeak")
    public void upgradeFirmware(final Uri uri, final boolean isItFirstAttempt) {
         new  AsyncTask<Void, Void, Boolean>() {
            @Override
            protected Boolean doInBackground(Void... voids) {
                try {
                    Log.e(TAG, "start AsyncTask upgradeFirmware");
                    return mAspenService.upgradeFirmware(FotaServiceContract.TYPE_DEVICE_HMD  , uri);
                } catch (Exception e) {
                    Log.e(TAG, "Exception on upgradeFirmware", e);
                }
                return false;
            }

            @Override
            protected void onPostExecute(Boolean result) {
                if (!result) {
                    myDelegate delegate = getDelegate();
                    if (delegate != null) {
                        delegate.onFotaError();
                    }
                }
            }
        }.executeOnExecutor(Executors.newFixedThreadPool(1));
    }

    public boolean isBinded() {
        return isBinded;
    }
}

File: aspen_fota/MIAC_ControllerUtility/src/com/htc/miac/controllerutility/utils/AspenServiceModel.kt
package com.htc.miac.controllerutility.utils

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.net.Uri
import android.os.*
import android.util.Log
import com.finchtechnologies.fota.IFotaListener
import com.finchtechnologies.fota.IFotaService
import java.util.concurrent.Executors

/**
 * Created by hubin_jiang on 2018/6/18.
 */
public class AspenServiceModel(val mContext: Context) {
    interface Delegate {
        fun onDeviceStatusChanged(state: Int, bundle: Bundle)
        fun onFirmwareUpdateProgressChanged(progress: Int)
        fun onDeviceInfoGet(info: Bundle?)
        fun onFotaError()
        fun onServiceConnected()
    }

    private val TAG = "AspenFota.client"

    val SERVICE_PACKAGE = "com.finchtechnologies.fota"
    private val SERVICE_CLASS = SERVICE_PACKAGE + ".FotaService"

    private var mIFinchFotaService: IFotaService? = null
    private val mUIHandler = Handler(Looper.getMainLooper())
    private var mDelegate: Delegate? = null
    private var mMacAddress = ""
    private var isBinded = false

    //private var mContext: Context;

//    fun AspenServiceModel(context: Context): ??? {
//        mContext = context
//    }

    fun bindService(): Boolean {
        Log.i(TAG, "bindService")
        var isBind = false
        try {
            val intent = Intent()
            intent.setClassName(SERVICE_PACKAGE, SERVICE_CLASS)
            isBind = mContext.bindService(intent, mConnection, Context.BIND_AUTO_CREATE)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return isBind
    }

    fun unbindService() {
        Log.d(TAG, "unbindService")
        mMacAddress = ""
        mContext.unbindService(mConnection)
        isBinded = false
    }

    fun setDelegate(delegate: Delegate) {
        mDelegate = delegate
    }

    fun getDelegate(): Delegate? {
        return mDelegate
    }

    private val mConnection = object : ServiceConnection {

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Log.d(TAG, "onServiceConnected")
            isBinded = true
            mIFinchFotaService = IFotaService.Stub.asInterface(service)
            try {
                mIFinchFotaService!!.setDeviceListener(mFinchFotaListener)
            } catch (e: RemoteException) {
                e.printStackTrace()
            }

            val delegate = getDelegate()
            delegate?.onServiceConnected() ?: Log.w(TAG, "[onServiceConnected] delegate is null !")
        }

        override fun onServiceDisconnected(name: ComponentName) {

        }
    }

    private val mFinchFotaListener = object : IFotaListener.Stub() {

        @Throws(RemoteException::class)
        override fun onDeviceStatusChanged(state: Int, extra: Bundle) {
            Log.d(TAG, "state : " + state)
            mUIHandler.post {
                val delegate = getDelegate()
                delegate?.onDeviceStatusChanged(state, extra)
            }
        }

        @Throws(RemoteException::class)
        override fun onFirmwareUpdateProgressChanged(progress: Int) {
            mUIHandler.post {
                val delegate = getDelegate()
                delegate?.onFirmwareUpdateProgressChanged(progress)
            }
        }
    }

    fun setMacAddress(addr: String) {
         object : AsyncTask<Void, Void, Void>() {
            override fun doInBackground(vararg voids: Void): Void? {
                try {
                    mIFinchFotaService!!.setMacAddress(addr)
                } catch (e: Exception) {
                    Log.e(TAG, "Exception on setMacAddress", e)
                }

                return null
            }
        }.executeOnExecutor(Executors.newFixedThreadPool(1))
    }

    fun getDeviceInfo() {
        object : AsyncTask<Void, Void, Bundle>() {

            override fun doInBackground(vararg voids: Void): Bundle? {
                try {
                    return mIFinchFotaService!!.deviceInfo
                } catch (e: Exception) {
                    Log.e(TAG, "Exception on getDeviceInfo", e)
                }

                return null
            }

            override fun onPostExecute(info: Bundle?) {
                val delegate = getDelegate()
                if (delegate != null) {
                    if (info != null) {
                        delegate.onDeviceInfoGet(info)
                    } else {
                        delegate.onFotaError()
                    }
                } else {
                    Log.e(TAG, "delegate is null")
                }
            }
        }.executeOnExecutor(Executors.newFixedThreadPool(1))
    }

    fun getBatteryVoltageLevel(): Int {
        var batteryLevel = -1
        try {
            batteryLevel = mIFinchFotaService!!.batteryVoltageLevel
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return batteryLevel
    }

    fun upgradeFirmware(uri: Uri, isItFirstAttempt: Boolean) {
        object : AsyncTask<Void, Void, Boolean>() {
            override fun doInBackground(vararg voids: Void): Boolean? {
                try {
                    return if (isItFirstAttempt) {
                        mIFinchFotaService!!.upgradeFirmware(uri)
                    } else {
                        mIFinchFotaService!!.upgradeFirmwareOnDfuTarg(uri)
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Exception on upgradeFirmware", e)
                }

                return false
            }

            override fun onPostExecute(result: Boolean) {
                if ( !result ) {
                    val delegate = getDelegate()
                    delegate?.onFotaError()
                }
            }
        }.executeOnExecutor(Executors.newFixedThreadPool(1))
    }

    fun isBinded(): Boolean {
        return isBinded
    }

    companion object{
        fun get_static_name(): String {
            return "static test"
        }
    }
}